pkg/reconciler/namespace/namespace.go

```diff
@@ -323,10 +370,16 @@ func (c *Controller) reconcileNamespace(ctx context.Context, lclusterName string
 	return nil
 }
 
-// clusterLabelPatchBytes returns JSON patch bytes expressing an operation to
-// add or replace the cluster assignment label to the given value.
-func clusterLabelPatchBytes(val string) []byte {
-	return []byte(fmt.Sprintf(`
+// ClusterLabelPatchBytes returns JSON patch bytes expressing an operation to
+// add, replace to the given value, or delete the cluster assignment label.
+func ClusterLabelPatchBytes(val string) (types.PatchType, []byte) {
```

Comment:
This seems to only be exported so that it can be called from e2e tests. If that's the case, I might suggest either reimplementing necessary bits in the test itself (i.e., test that the patch is `"[{"op":...}]` instead of reusing the helper), or perhaps moving the method to an `/internal/` package so that it can't be depended on outside of this Go module.

Exported methods have a nasty habit of getting depended on outside of our codebase, and sometimes beyond intended usage, which can make changes harder to make safely.

Reply:
I agree. @marun what do you think of moving that function in the `/internal/` package? Could that be an acceptable compromise, as you raised concerns about duplicating it?

Reply:
Agreed that this detail shouldn't be exposed. I don't think jumping through the hoops of moving to internal is suggested though. How about we avoid using low-level patch semantics in testing and rely on `retry.RetryOnConflict` instead?

e.g. https://github.com/kcp-dev/kcp/blob/main/test/e2e/reconciler/ingress/controller_test.go#L102